maps-compose/src/main/java/com/google/maps/android/compose/GoogleMap.kt

```diff
@@ -123,115 +124,116 @@ public fun GoogleMap(
-    val lifecycle = LocalLifecycleOwner.current.lifecycle
-    val previousState = remember { mutableStateOf(Lifecycle.Event.ON_CREATE) }
-    DisposableEffect(context, lifecycle, mapView) {
-        val mapLifecycleObserver = mapView.lifecycleObserver(previousState)
-        val callbacks = mapView.componentCallbacks()
-
-        lifecycle.addObserver(mapLifecycleObserver)
-        context.registerComponentCallbacks(callbacks)
-
-        onDispose {
-            lifecycle.removeObserver(mapLifecycleObserver)
-            context.unregisterComponentCallbacks(callbacks)
+            CompositionLocalProvider(
+                LocalCameraPositionState provides currentCameraPositionState,
+            ) {
+                currentContent?.invoke()
+            }
         }
-    }
-    DisposableEffect(mapView) {
-        onDispose {
-            mapView.onDestroy()
-            mapView.removeAllViews()
+
+        return launch(start = CoroutineStart.UNDISPATCHED) {
+            val map = mapView.awaitMap()
+            val composition = Composition(
+                applier = MapApplier(map, mapView, mapClickListeners),
+                parent = parentComposition
+            )
+            composition.setContent(mapCompositionContent)
+
+            try {
+                awaitCancellation()
+            } finally {
+                composition.dispose()
+            }
         }
     }
-}
 
-private fun MapView.lifecycleObserver(previousState: MutableState<Lifecycle.Event>): LifecycleEventObserver =
-    LifecycleEventObserver { _, event ->
-        event.targetState
-        when (event) {
-            Lifecycle.Event.ON_CREATE -> {
-                // Skip calling mapView.onCreate if the lifecycle did not go through onDestroy - in
-                // this case the GoogleMap composable also doesn't leave the composition. So,
-                // recreating the map does not restore state properly which must be avoided.
-                if (previousState.value != Lifecycle.Event.ON_STOP) {
-                    this.onCreate(Bundle())
+    var subcompositionJob by remember { mutableStateOf<Job?>(null) }
+    val parentCompositionScope = rememberCoroutineScope()
+
+    AndroidView(
+        modifier = modifier,
+        factory = { context ->
+            MapView(context, googleMapOptionsFactory()).also { mapView ->
+                // Register ComponentCallbacks so that MapView is notified on low memory
+                val componentCallbacks = mapView.registerComponentCallbacks()
+                // Forward parent and custom lifecycle events to MapView
+                val lifecycleObserver = MapLifecycleEventObserver(mapView)
+                // Store things in the tag which must be retrievable across recompositions
+                mapView.tag = MapTagData(componentCallbacks, lifecycleObserver)
+                // MapView's parent LifecycleOwner, such as Activity/Fragment
+                var lifecycleOwner: LifecycleOwner? = null
```

Comment:
I'm not entirely sure what you mean. Do you mean to add it to object : View.OnAttachStateChangeListener as a member? Because that would make sense.

Reply:
> I'm not entirely sure what you mean. Do you mean to add it to `object : View.OnAttachStateChangeListener` as a member? Because that would make sense.

Yes, that would be ideal. I meant store `lifecycle` instead of `lifecycleOwner`. Previously we zeroed in on `lifecycleOwner` because we received it from composition, but here we really only need to care about the `lifecycle` object specifically.

Reply:
True. We can store the `Lifecycle` instead